### PR TITLE
chore(shorebird_code_push_protocol): Add metadata messages

### PR DESCRIPTION
**Pull request by bryanoltman**
Adds metadata classes for recording information about release and patch command runs. Will be used to more effectively diagnose release and patch issues and to better understand how shorebird is being used.

### CODE REVIEW

File: packages/shorebird_code_push_protocol/lib/src/models/build_environment_metadata.dart [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+  /// Converts a Map<String, dynamic> to a [BuildEnvironmentMetadata]
+  factory BuildEnvironmentMetadata.fromJson(Map<String, dynamic> json) =>
+      _$BuildEnvironmentMetadataFromJson(json);
+
+  /// Converts a [BuildEnvironmentMetadata] to a Map<String, dynamic>
+  Map<String, dynamic> toJson() => _$BuildEnvironmentMetadataToJson(this);
+
+  /// The version of Shorebird used to run the command.
+  final String shorebirdVersion;
```
Comment (eseidel) [inline packages/shorebird_code_push_protocol/lib/src/models/build_environment_metadata.dart]: Can you explain the reason next to each, e.g.:

```suggestion
  /// The version of Shorebird used to run the command.
  /// This is stored with the release and displayed in console.shorebird.dev.
  final String shorebirdVersion;
```

File: packages/shorebird_code_push_protocol/lib/src/models/build_environment_metadata.dart [hunk 2]
```diff
@@ -0,0 +1,63 @@
+import 'package:equatable/equatable.dart';
+import 'package:json_annotation/json_annotation.dart';
+
+part 'build_environment_metadata.g.dart';
+
+/// {@template build_environment_metadata}
+/// Information about the environment used to build a release or patch.
+/// {@endtemplate}
```
Comment (eseidel) [inline packages/shorebird_code_push_protocol/lib/src/models/build_environment_metadata.dart]:
```suggestion
/// Information about the environment used to build a release or patch.
/// Collection of this build environment information is done to help Shorebird
/// users debug any later failures in their builds.
/// We do not collect Personally Identifying Information (e.g. no paths, argument lists, etc.)
/// in accordance with our privacy policy: https://shorebird.dev/privacy/
/// {@endtemplate}
```

**Review by eseidel — APPROVED** [review]
lgtm.  In general I'd like to us to take a posture of being over cautious and over-explaining on what data we send from our users machines to our servers.  Please add individual comments about the fields being collected if possible.  Thank you.

File: packages/shorebird_code_push_protocol/lib/src/models/build_environment_metadata.dart (hunk 2 above)
Comment (bryanoltman) [reply packages/shorebird_code_push_protocol/lib/src/models/build_environment_metadata.dart]: Done

File: packages/shorebird_code_push_protocol/lib/src/models/build_environment_metadata.dart (hunk 1 above)
Comment (bryanoltman) [reply packages/shorebird_code_push_protocol/lib/src/models/build_environment_metadata.dart]: Done

**Pull request merged**